README.md
# **LightningStrike ⚡**

The zero config, self-managed, bitcoin lightning marketplace.

Take back ownership of your market!

  * [BTC Lightning payment](https://lightning.network/)
  * [Strike API integration](https://developer.strike.me/en/)
  * [Tor routed .onion hostname](https://www.torproject.org/)

![1_home.png](/screenshoots/1_home.png)

![2_home.png](/screenshoots/2_pay_with_lightning.png)

![3_home.png](/screenshoots/3_wait_for_payment.png)

![4_home.png](/screenshoots/4_thankyou.png)


## Startup

Requires [Docker](https://www.docker.com/get-started)

```
$ export ENABLE_TOR=true  # optional
$ export TOKEN="YOUR STRIKE API TOKEN"
$ docker-compose up
```

If you don't yet have a strike api token, you can still explore the platform;
merely unset your $TOKEN variable before run:

```
$ unset TOKEN
$ docker-compose up
```

Now, when you press the "pay with lightning" button, we will respond with mock data (id is all 0s).

You should be able to access http://127.0.0.1:8080


## Inventory

The **inventory.yaml** file is the heart of your marketplace.

The structure of this file should be simple enough for any user to manage,
check out [YAML](https://yaml.org/) if you need more detail.

### Images (thumbnails)

Our inventory file contains a **key** called `images` for each product:

*inventory.yaml*
```
- name: Pancakes
  description: More like tasty cakes
  price: 0.05
  image: pancakes.jpeg
  content: pancakes.jpeg
```

These images are hosted within the `static/images/` folder; for example,
**pancakes.jpeg** can be found at `static/images/pancakes.jpeg`


### Content

Our inventory file also has a **key** called `content` for each product,
as expected, these live in `content/` and are only accessible through a valid
receipt page.

A receipt page is only valid when the following are true:

  1. invoice's state is **PAID**
  2. invoice created less than **24 hours** ago.


## Banner

The banner can be changed by replacing `static/images/banner.jpeg` with your own.

## Templates

The look and feel of your application lives within your templates, more
accurately, your `templates/` folder. You have the ability to change any
aspect of your marketplace and the full power of the
[Jinja2](https://jinja2docs.readthedocs.io/en/stable/templates.html) template
renderer.


## Sales

Without adding much complexity, or more compute resources in the form of
database servers, I decided to go with simple comma-separated values file.

As **invoices** are created, and as invoice are **paid** your `sales.csv` will grow.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.0.1] - 2022-02-22
- A generic, easy to setup, online market place.
- [Bitcoin lightning](https://lightning.engineering/) payments using [Strike API](https://strike.me).
- Optionally [Tor](https://www.torproject.org/) routed (privacy and/or zero domain config).


utils/strike.py
#!/usr/bin/env python3

import sys
import os
import time
import uuid
from datetime import datetime, timezone

import json
import requests


# Strike api documentation
# docs.strike.me
#
# Strike api token should be set as a env variable
TOKEN = os.environ['TOKEN']

dry_run=False
if not TOKEN:
    dry_run=True


# Basic auth headers for strike api
HEADERS = {
  'Content-Type': 'application/json',
  'Accept': 'application/json',
  'Authorization': 'Bearer %s' % TOKEN
}


def create_invoice(description, amount):
    """
    Call strike's restful api to create a new invoice
    """

    cor_id = uuid.uuid4().hex
    url = "https://api.strike.me/v1/invoices"

    payload = json.dumps({
      "correlationId": cor_id,
      "description": description,
      "amount": {
        "currency": "USD",
        "amount": amount
      }
    })

    if dry_run:
        return {
            'created': str(datetime.utcnow().replace(tzinfo=timezone.utc)),
            'invoice_id': '000000000-0000-0000-0000-000000000000',
            'cor_id': cor_id,
            'description': description,
            'amount': amount
        }
    else:
        response = requests.request("POST", url, headers=HEADERS, data=payload)
        response.raise_for_status()
        rj = response.json()

        created = rj['created']
        invoice_id = rj['invoiceId']
        description = rj['description']
        price = rj['amount']['amount']
        state = 'CREATED'

        return {
            'created': created,
            'invoice_id': invoice_id,
            'cor_id': rj['correlationId'],
            'description': description,
            'amount': price
        }



def get_invoice(invoice_id):
    """
    Call strike's restful api to get an invoice by id
    """

    url = "https://api.strike.me/v1/invoices/%s" % invoice_id
    payload={}

    if dry_run:
        return {
            'created': str(datetime.utcnow().replace(tzinfo=timezone.utc)),
            'invoice_id': '000000000-0000-0000-0000-000000000000',
            'cor_id': 'b3c540adfe3b4185b481ae1a467afbeb',
            'description': 'Coffee',
            'amount': '0.01',
            'state': 'PAID'
        }
    else:
        response = requests.request("GET", url, headers=HEADERS, data=payload)
        response.raise_for_status()

        rj = response.json()

        created = rj['created']
        invoice_id = rj['invoiceId']
        description = rj['description']
        price = rj['amount']['amount']
        state = rj['state']

        return {
            'created': created,
            'invoice_id': invoice_id,
            'cor_id': rj['correlationId'],
            'description': description,
            'amount': price,
            'state': state
        }


def create_ln_invoice(invoice_id):
    """
    Call strike's restful api to create a new lightning quote
    """

    url = "https://api.strike.me/v1/invoices/%s/quote" % invoice_id
    payload={}

    if dry_run:
        return {
            'ln_invoice': 'lnbc1pvjluezsp5zyg3zyg3zyg3zyg3zyg3zyg3zyg3zyg3zyg3zyg3zyg3zyg3zygspp5qqqsyqcyq5rqwzqfqqqsyqcyq5rqwzqfqqqsyqcyq5rqwzqfqypqdpl2pkx2ctnv5sxxmmwwd5kgetjypeh2ursdae8g6twvus8g6rfwvs8qun0dfjkxaq9qrsgq357wnc5r2ueh7ck6q93dj32dlqnls087fxdwk8qakdyafkq3yap9us6v52vjjsrvywa6rt52cm9r9zqt8r2t7mlcwspyetp5h2tztugp9lfyql',
            'expiration_in_sec': 117
        }
    else:
        response = requests.request("POST", url, headers=HEADERS, data=payload)
        response.raise_for_status()

        rj = response.json()
        return {
            'ln_invoice': rj['lnInvoice'],
            'expiration_in_sec': rj['expirationInSec']
        }


def create_qr(ln_invoice):
    """
    Generate a base64 encoded QR code of our lightning invoice
    """

    qr = QRCode()
    qr.add_data(ln_invoice)

    buffered = BytesIO()
    img = qr.make_image()
    img.save(buffered, format="PNG")

    buffered.seek(0)
    img_byte = buffered.getvalue()
    img_str = "data:image/png;base64," + base64.b64encode(img_byte).decode()

    return img_str


utils/csv_tools.py
import os

import csv


FILENAME = "sales.csv"
FIELDS = ['created', 'invoice_id', 'description', 'price', 'state']


def create():
    """
    Create the sales.csv
    """

    with open(FILENAME, 'w', newline='') as csvfile:
        csvwriter = csv.writer(csvfile)
        csvwriter.writerow(FIELDS)


def append(created, invoice_id, description, price, state):
    """
    Append to the the sales.csv
    """

    with open(FILENAME, 'a', newline='') as csvfile:
        csvwriter = csv.writer(csvfile)
        csvwriter.writerow([created, invoice_id, description, price, state])


def create_or_append(created, invoice_id, description, price, state):
    """
    Create or append to our sales.csv
    """

    if not os.path.exists('sales.csv'):
        create()

    append(created, invoice_id, description, price, state)


static/js/scripts.js
var myTimeout = null;

function createInvoice(description, price) {

  var http = new XMLHttpRequest();
  var url = '/invoice/';
  var params = 'description=' + description + '&price=' + price;

  http.open('POST', url , true);
  http.setRequestHeader('Content-type', 'application/x-www-form-urlencoded');
  http.send(params);

  http.onreadystatechange = function() {
    if (this.status === 200 && this.readyState === 4) {
      var res = JSON.parse(http.response);
      console.log(res);
      createLightningInvoice(res['invoice_id']);
    }
  }

}

function createLightningInvoice(invoice_id) {

  var http = new XMLHttpRequest();
  var url = '/ln_invoice/';
  var params = 'invoice_id=' + invoice_id;

  http.open('POST', url , true);
  http.setRequestHeader('Content-type', 'application/x-www-form-urlencoded');
  http.send(params);

  http.onreadystatechange = function() {
    if (this.status === 200 && this.readyState === 4) {
      var res = JSON.parse(http.response);
      console.log(res);

      // update button
      document.getElementById('buy-button').style.display = 'none'

      document.querySelector('#Modal' + ' .qr-content').style.display = 'block'
      document.getElementById('invoice_id').innerHTML = 'id: ' + invoice_id;
      document.getElementById('ln_invoice_qr').src = res['ln_invoice_qr'];
      document.getElementById('ln_invoice').innerHTML = res['ln_invoice'];
      document.getElementById('ln_timer').innerHTML = res['expiration_in_sec'];
      checkInvoice(invoice_id);

    }
  }

}

function getInvoice(invoice_id) {

  var http = new XMLHttpRequest();
  var url = '/invoice/' + invoice_id;

  http.open('GET', url , true);
  http.setRequestHeader('Content-type', 'application/x-www-form-urlencoded');
  http.send();

  http.onreadystatechange = function() {
    if (this.status === 200 && this.readyState === 4) {
      var res = JSON.parse(http.response);
      console.log(res);

      if (res['state'] == 'PAID') {
        window.location.replace("/receipt/" + invoice_id);
      }
    }
  }

}

function checkInvoice(invoice_id) {

  getInvoice(invoice_id);
  myTimeout = setTimeout(function(){
    checkInvoice(invoice_id);
  }, 5000);

}



function closeModal() {
  clearInvoice()
  // update button
  document.getElementById('buy-button').style.display = 'inline-block'
  document.querySelector('#Modal' + ' .qr-content').style.display = 'none'
}

function clearInvoice() {
  document.getElementById('invoice_id').innerHTML = '';
  document.getElementById('ln_invoice_qr').src = '';
  document.getElementById('ln_invoice').innerHTML = '';
  document.getElementById('ln_timer').innerHTML = '';

  clearTimeout(myTimeout);
  myTimeout = null;
}


app.py
import os
from datetime import datetime, timezone
from dateutil import parser

import yaml
from flask import Flask, render_template, request, abort, send_file

from slugify import slugify
from utils.csv_tools import create_or_append
from utils.qr import create_qr
from utils import strike

app = Flask(__name__)



def get_items(items_path='inventory.yaml'):
    """
    Get items from yaml file
    """

    items = []
    if os.path.exists(items_path):
        with open(items_path) as f:
            items = yaml.safe_load(f.read())

    # add item slugs
    for item in items:
        item['slug'] = slugify(item['name'])

    return items


@app.route("/")
def home():
    """
    Render and return templates/home.html
    """

    # Get list of items
    items = get_items()

    return render_template('home.html', items=items)


@app.route("/product/<slug_name>")
def product(slug_name):
    """
    Render and return templates/product.html
    """

    # Get list of items
    items = get_items()

    for item in items:
        if item['slug'] == slug_name:
            return render_template('product.html', item=item)

    return abort(404)


@app.route("/receipt/<invoice_id>")
def receipt(invoice_id):
    """
    Render and return templates/receipt.html only if invoice is paid
    """

    # Get list of items
    # TODO add func
    items = []
    items_path = 'inventory.yaml'
    if os.path.exists(items_path):
        with open(items_path) as f:
            items = yaml.safe_load(f.read())

    invoice = strike.get_invoice(invoice_id)

    if invoice['state'] == 'PAID':

        # only show receipts for 24 hours
        created = parser.parse(invoice['created'])
        delta = datetime.utcnow().replace(tzinfo=timezone.utc) - created

        if delta.days <= 1:

            # get origin item using description...
            # yuck, but no database, so it will do
            for item in items:
                if item['name'] == invoice['description']:

                    return render_template(
                        'receipt.html', invoice_id=invoice_id)

    # return not found
    return abort(404)


@app.route("/download/<invoice_id>")
def download(invoice_id):
    """
    Pass file to user if their invoice is valid
    """

    # Get list of items
    # TODO add func
    items = []
    items_path = 'inventory.yaml'
    if os.path.exists(items_path):
        with open(items_path) as f:
            items = yaml.safe_load(f.read())

    invoice = strike.get_invoice(invoice_id)

    if invoice['state'] == 'PAID':

        # only show receipts for 24 hours
        created = parser.parse(invoice['created'])
        delta = datetime.utcnow().replace(tzinfo=timezone.utc) - created

        if delta.days <= 1:

            # get origin item using description...
            # yuck, but no database, so it will do
            for item in items:
                if item['name'] == invoice['description']:

                    content_path = 'content/%s' % item['content']

                    return send_file(content_path)

    # return not found
    return abort(404)



@app.route("/about")
def about():
    """
    Render and return templates/about.html
    """

    return render_template('about.html')


@app.route("/invoice/", methods=['POST'])
def create_invoice():
    """
    Create strike invoice
    """

    description = request.form['description']
    price = request.form['price']

    invoice = strike.create_invoice(description, price)

    create_or_append(
        invoice['created'], invoice['invoice_id'], invoice['description'],
        invoice['amount'], 'CREATED'
    )

    return invoice


@app.route("/ln_invoice/", methods=['POST'])
def create_ln_invoice():
    """
    Create strike lightning quote
    """

    invoice_id = request.form['invoice_id']

    ln_invoice = strike.create_ln_invoice(invoice_id)

    # encode as qrcode
    ln_invoice['ln_invoice_qr'] = create_qr(ln_invoice['ln_invoice'])

    return ln_invoice


@app.route("/invoice/<invoice_id>")
def get_invoice(invoice_id):
    """
    Get strike invoice
    """

    invoice = strike.get_invoice(invoice_id)

    if invoice['state'] == 'PAID':
        create_or_append(
            invoice['created'], invoice['invoice_id'], invoice['description'],
            invoice['amount'], invoice['state']
        )

    return invoice


app.run(host='0.0.0.0', port='8080', debug=False)
